Feature/session controller api

dkattan opened the pull request:
Added APIs that allow for creating public URLs to access remote sessions.

---

Please read the following.  Do not delete below this line.

---

Thank you for your contribution to the Remotely project.  It is required that contributors assign copyright to Immense Networks to retain full ownership of the project.

This makes it easier for other entities to use the software because they only have to deal with one copyright holder.  It also gives me assurance that I'll be able to make decisions in the future without gathering and consulting all contributors.

While this may seem odd, many open source maintainers practice this.  Here are a couple well-known examples:

- Free Software Foundation: http://www.gnu.org/licenses/why-assign.html
- Microsoft: https://cla.opensource.microsoft.com/

A nice article on the topic can be found here:  https://haacked.com/archive/2006/01/26/WhoOwnstheCopyrightforAnOpenSourceProject.aspx/

By submitting this PR, you agree to the following:

> You hereby assign copyright in this PR's code to the Remotely project and its copyright holder, Immense Networks to be licensed under the same terms as the rest of the code.  You agree to relinquish any and all copyright interest in the software, to the detriment of your heirs and successors.

lucent-sea commented on Server/API/RemoteControlController.cs:
```diff
@@ -77,12 +77,19 @@ public async Task<IActionResult> Post([FromBody] RemoteControlRequest rcRequest)
+                    .Where(x => x.Value.DeviceID == deviceID)
+                    .Select(x => x.Key)
+                    .ToList();
+            if (existingSessions == null)
```
This would never be null.  So we'd want `(!existingSessions.Any())`.

lucent-sea commented on Server/API/RemoteControlController.cs:
```diff
@@ -114,17 +118,17 @@ bool remoteControlStarted()
                 {
                     return StatusCode(408, "The remote control process failed to start in time on the remote device.");
                 }
```
If we successfully start a remote control session here, we're not doing anything with it.  We drop below the if/else and return BadRequest.

lucent-sea commented on Server/API/RemoteControlController.cs:
```diff
@@ -114,17 +118,17 @@ bool remoteControlStarted()
+                    return Ok($"{HttpContext.Request.Scheme}://{Request.Host}/RemoteControl?casterID={rcSession.CasterSocketID}&serviceID={rcSession.ServiceID}&fromApi=true&otp={Uri.EscapeDataString(otp)}");
+                }
             }
+            return BadRequest("The target device couldn't be found.");
```
Again, NotFound?  :)

lucent-sea commented on Server/API/RemoteControlController.cs:
```diff
@@ -77,12 +77,19 @@ public async Task<IActionResult> Post([FromBody] RemoteControlRequest rcRequest)
+                                        x.Value.ID.ToLower() == deviceID.ToLower());
+                if (targetDevice.Value is null)
+                {
+                    return BadRequest("The target device couldn't be found.");
```
Pedantic, but should this be NotFound (404)?  :)

lucent-sea reviewed (changes requested):
Didn't have time to look at ViewerHub, but wanted to write down my comments.

Aside from the comments, am I understanding correctly that the purpose of these changes is to avoid launching a new remote control session and join an existing one if it's already started?

Just FYI, it can have multiple remote control sessions (e.g. Remotely_Desktop.exe processes) going at once on the same devce.  Or should be able to.  I thought it would be best to keep them in separate processes when possible.  But maybe it's better for your workflow to have them using the same process?

lucent-sea commented on Server/API/RemoteControlController.cs:
```diff
@@ -114,17 +118,17 @@ bool remoteControlStarted()
+                var rcSession = CasterHub.SessionInfoList.Values.LastOrDefault(x => x.DeviceID == deviceID);
+                if (rcSession != null)
+                {
+                    var otp = RemoteControlFilterAttribute.GetOtp(deviceID);
```
We'd want to verify that user has authorization to access device with `DataService.DoesUserHaveAccessToDevice`.

dkattan commented:
> Didn't have time to look at ViewerHub, but wanted to write down my comments.
> 
> Aside from the comments, am I understanding correctly that the purpose of these changes is to avoid launching a new remote control session and join an existing one if it's already started?
> 
> Just FYI, it can have multiple remote control sessions (e.g. Remotely_Desktop.exe processes) going at once on the same devce. Or should be able to. I thought it would be best to keep them in separate processes when possible. But maybe it's better for your workflow to have them using the same process?



> Didn't have time to look at ViewerHub, but wanted to write down my comments.
> 
> Aside from the comments, am I understanding correctly that the purpose of these changes is to avoid launching a new remote control session and join an existing one if it's already started?
> 
> Just FYI, it can have multiple remote control sessions (e.g. Remotely_Desktop.exe processes) going at once on the same devce. Or should be able to. I thought it would be best to keep them in separate processes when possible. But maybe it's better for your workflow to have them using the same process?

The core purpose is to allow an integration to get a public link (with an OTP) to a session for a machine that doesn't exist in Remotely. Let's call this an "ad-hoc" session.

Here's the code that uses this API (assume the integration has the ability to run PowerShell on the target device)

```csharp
    public async Task<IScript> GetAdhocPowerShellInstallScript(GetPowerShellAdhocSessionScriptParameters requestOptions, CancellationToken token)
    {
      // DeviceId here is the integration's deviceId. It may be better to let Remotely generate a deviceid to prevent potential collisions
      if (string.IsNullOrEmpty(requestOptions.DeviceId))
        return null;

      var organizationId = requestOptions.RmmClient.RmmClientId.Split("/").FirstOrDefault();
      var deviceGroup = requestOptions.RmmClient.RmmClientId.Split("/")[1];      
      var installerUri = new Uri($"{_formData?.RemotelyApiUri}api/ClientDownloads/Desktop/WindowsDesktop-x86/{organizationId}");
      var script = new Script();
      script.ScriptActionType = ScriptActionType.PowerShell;
      script.Timeout = 300;
      var authHeader = _httpClient.DefaultRequestHeaders.FirstOrDefault(a => a.Key == "Authorization");
      if (authHeader.Value is not null)
      {
        var value = authHeader.Value.FirstOrDefault();
        var headers = new Dictionary<string, string>();
        headers.Add(authHeader.Key, value);
        script.Action = ScriptBuilder.BuildPowerShellDownloadAndInstallScript(installerUri, $@"-mode Unattended -host ""{_formData?.RemotelyApiUri}"" -deviceid ""{requestOptions.DeviceId}""", headers);
        script.ScriptExecutionContext = ScriptExecutionContext.CurrentUser;      }
      return script;
    }
    // The ImmyBot user is redirected to the URL returned here
    public async Task<Uri> GetAdhocRemoteUrl(string deviceId, string userDisplayName, bool requireConsent, CancellationToken token)
    {
      try
      {
        var url = await _httpClient.GetStringAsync($"/api/RemoteControl/{deviceId}");
        return new Uri(url);
      }
      catch (Exception ex)
      {
        return null;
      }
    }
  }
```

lucent-sea commented:
In that case, I'm confused.  Didn't the original version already give you an OTP code in the URL?

I'd recommend keeping it so it always launches a new process/session.  I realized that if there's an existing session, and you return a URL to that existing one, the viewer might leave and close the session before you are able to connect.

dkattan commented:
> In that case, I'm confused. Didn't the original version already give you an OTP code in the URL?
> 
> I'd recommend keeping it so it always launches a new process/session. I realized that if there's an existing session, and you return a URL to that existing one, the viewer might leave and close the session before you are able to connect.

Yes, but it fails if the device doesn't exist in the AgentHub. The operative change here is to provide a code path that works if the deviceid isn't found in the AgentHub.

lucent-sea commented:
Oh yeah!  I forgot that you didn't have the agent installed.  My bad.   :D

Then yeah, I'd just recommend adding an optional bool param to that private method, indicating it should look for existing sessions instead of launching a new process.  Then create a separate GET endpoint for it.

That way, the existing logic isn't altered in any way that could possibly break existing integrations.
